(chore) set min version of cmake in native CmakeLists.txt to surpass Cmake CMP0000 policy warning

Neo-vortex opened the pull request:
This closes https://github.com/dotnet/runtime/issues/93436

Setting the min version for `cmake` reflecting `https://github.com/dotnet/runtime/blob/main/src/native/libs/CMakeLists.txt`

Neo-vortex committed:
(chore) set min version of cmake in native CmakeLists.txt to surpass policy CMP0000 of Cmake

Neo-vortex commented:
@dotnet-policy-service agree

Neo-vortex committed:
(chore) set all cmake versions according to native/libs/CMakeLists.txt

ViktorHofer commented:
cc @jkoritzinsky as you dealt with minimum cmake versions in our repo in the past

jkotas commented on src/native/libs/System.Globalization.Native/CMakeLists.txt:
```diff
@@ -1,3 +1,5 @@
+cmake_minimum_required(VERSION 3.20)
```
```suggestion
# Required for StaticICULinking
cmake_minimum_required(VERSION 3.20)
```

jkotas commented on src/native/libs/System.IO.Compression.Native/CMakeLists.txt:
```diff
@@ -1,3 +1,5 @@
+cmake_minimum_required(VERSION 3.20)
+
```
```suggestion
```
We need this only for the libraries that we support static linking for. The fewer places with cmake version, the better.

jkotas commented on src/native/libs/System.IO.Ports.Native/CMakeLists.txt:
```diff
@@ -1,3 +1,5 @@
+cmake_minimum_required(VERSION 3.20)
+
```
```suggestion
```

jkotas commented on src/native/libs/System.Security.Cryptography.Native/CMakeLists.txt:
```diff
@@ -1,3 +1,5 @@
+cmake_minimum_required(VERSION 3.20)
```
```suggestion
# Required for StaticOpenSslLinking
cmake_minimum_required(VERSION 3.20)
```

jkotas commented on src/native/libs/System.Native/CMakeLists.txt:
```diff
@@ -1,3 +1,5 @@
+cmake_minimum_required(VERSION 3.20)
+
```
```suggestion
```

jkotas commented on src/native/libs/System.Net.Security.Native/CMakeLists.txt:
```diff
@@ -1,3 +1,5 @@
+cmake_minimum_required(VERSION 3.20)
+
```
```suggestion
```

jkotas commented on src/native/libs/System.Security.Cryptography.Native.Android/CMakeLists.txt:
```diff
@@ -1,3 +1,5 @@
+cmake_minimum_required(VERSION 3.20)
+
```
```suggestion
```

jkotas commented on src/native/libs/System.Security.Cryptography.Native.Apple/CMakeLists.txt:
```diff
@@ -1,3 +1,5 @@
+cmake_minimum_required(VERSION 3.20)
+
```
```suggestion
```

jkotas committed:
Apply suggestions from code review

jkotas committed:
Update src/native/libs/System.Security.Cryptography.Native/CMakeLists.txt

jkotas reviewed (approved):
Thank you

jkoritzinsky commented on src/native/libs/System.Globalization.Native/CMakeLists.txt:
```diff
@@ -1,3 +1,6 @@
+# Required for StaticICULinking
+cmake_minimum_required(VERSION 3.20)
```
```suggestion
cmake_minimum_required(VERSION 3.16...3.20)
```

Can we change the minimum version to 3.16 to be consistent with other CMake scripts we ship to external customers?

jkotas replied:
What are the other CMake scripts that we ship?

jkoritzinsky commented on src/native/libs/System.Security.Cryptography.Native/CMakeLists.txt:
```diff
@@ -1,3 +1,6 @@
+# Required for StaticOpenSslLinking
+cmake_minimum_required(VERSION 3.20)
```
```suggestion
cmake_minimum_required(VERSION 3.16...3.20)
```

Neo-vortex committed:
Update src/native/libs/System.Globalization.Native/CMakeLists.txt

hange the minimum version to 3.16 to be consistent with other CMake scripts for System.Globalization.Native/CMakeLists.txt

Co-authored-by: Jeremy Koritzinsky <jkoritzinsky@gmail.com>

Neo-vortex committed:
Update src/native/libs/System.Security.Cryptography.Native/CMakeLists.txt

change the minimum version to 3.16 to be consistent with other CMake scripts 
 for System.Globalization.Native/CMakeLists.txt

Co-authored-by: Jeremy Koritzinsky <jkoritzinsky@gmail.com>

am11 commented on src/native/libs/System.Globalization.Native/CMakeLists.txt:
```diff
@@ -1,3 +1,6 @@
+# Required for StaticICULinking
+cmake_minimum_required(VERSION 3.16...3.20)
```
What happens when we have > 3.20 installed? Do we want to restrict on max value or can it just be set to 3.16 minimum?

Neo-vortex replied:
you are right.
limiting the max version is probably not what we want to do.
we can set `cmake_minimum_required(VERSION 3.16)`
fixed in the latest commit

jkoritzinsky replied:
This is not setting a max version. This is setting the version of CMake to set the policy values for. Setting it to 3.20 ensures that for the dotnet/runtime build, we use a consistent set of CMake policy configurations.

Neo-vortex committed:
(chore) set min version for cmake to 3.16

jkotas commented:
How did you come up with 3.16?

It would be better for this version to match the min cmake version used in the rest of the repo.

am11 commented:
> How did you come up with 3.16?

@jkotas, please expand the comments to see the history. It was a comment from @jkoritzinsky to use the version range 3.16...3.20, on which I commented that maybe we don't need to restrict to the max value. I think some old distros are providing older versions, so it makes sense to lower the requirements of "shipping" cmake script, which are not using any new feature.

jkoritzinsky commented:
3.16 is the minimum required version for the CMake scripts we use to build Apple and Android apps on customer machines (see the CMake templates in AppleAppBuilder and AndroidAppBuilder).

Neo-vortex commented:
@jkotas  
if @jkoritzinsky is right about the min cmake version, probably when can merge this PR as is.

jkotas commented:
I do not see a strong reason for keeping the Apple/Android builder min cmake version to be in sync with min cmake version for this Linux-specific opt-in option.

I see the following options:

- Use `3.20` to keep things simple. 
- Use `3.16..3.20`. As @jkoritzinsky pointed out, it is important to set 3.20 as the upper bound so that the dotnet/runtime build uses consistent set of policies. 
- Match cmake version that comes with the oldest Linux distro supported by .NET runtime. If it is going [`Ubuntu 18.04`](https://github.com/dotnet/runtime/issues/91826#issuecomment-1714378388) for .NET 9, it would make it `3.10..3.20`.

am11 commented:
>  it is important to set 3.20 as the upper bound so that the dotnet/runtime build uses consistent set of policies.

We don't have any restriction on > 3.20 version. In fact, the _minimum_ version is 3.20.
https://github.com/dotnet/runtime/blob/765798c49d5d908e512af0931b18fed782b4cf00/src/coreclr/CMakeLists.txt#L1

jkoritzinsky commented:
> > it is important to set 3.20 as the upper bound so that the dotnet/runtime build uses consistent set of policies.
> 
> We don't have any restriction on > 3.20 version. In fact, the _minimum_ version is 3.20.
> 
> https://github.com/dotnet/runtime/blob/765798c49d5d908e512af0931b18fed782b4cf00/src/coreclr/CMakeLists.txt#L1

One aspect of setting the minimum required as 3.20 in that was is to set the CMake policy defaults to the 3.20 rules, no matter the version of CMake that’s actually running the build. Using the `3.10…3.20` syntax says “3.10 is the minimum required version for the features used. Use the newest version of CMake policies available until 3.20. Then use the 3.20 policy defaults”

Neo-vortex commented:
> > > it is important to set 3.20 as the upper bound so that the dotnet/runtime build uses consistent set of policies.
> > 
> > 
> > We don't have any restriction on > 3.20 version. In fact, the _minimum_ version is 3.20.
> > https://github.com/dotnet/runtime/blob/765798c49d5d908e512af0931b18fed782b4cf00/src/coreclr/CMakeLists.txt#L1
> 
> One aspect of setting the minimum required as 3.20 in that was is to set the CMake policy defaults to the 3.20 rules, no matter the version of CMake that’s actually running the build. Using the `3.10…3.20` syntax says “3.10 is the minimum required version for the features used. Use the newest version of CMake policies available until 3.20. Then use the 3.20 policy defaults”

so `3.10..3.20` is the the best practice here where :
1- builds fail on anything below `3.10`
2- anything above `3.20` would use `3.20` policy defaults

Neo-vortex committed:
(chore) set min cmae version to 3.10
